BehaviorTreeEditor.cs

```diff
@@ -9,276 +9,391 @@
 
-		++_indentLevel;
-		
-		if ( node is Decorator )
-		{
-			height = CreateDecoratorGUI( (Decorator)node, height );
+				// handle drawing the child
+				DrawNodeCurve( nodeData.nodeRect, childData.nodeRect );
+				DrawNode( decorator._child );
+			}
+			else if ( node is Compositor )
+			{
+				Compositor compositor = (Compositor)node;
+
+				foreach ( TreeNode child in compositor._children )
+				{
+					EditorData childData = _editorData[child.id];
+
+					// handle drawing child
+					DrawNodeCurve( nodeData.nodeRect, childData.nodeRect );
+					DrawNode( child );
+				}
+			}
 		}
-		else if ( node is Compositor )
+
+		static string DisplayName( string name )
 		{
-			height = CreateCompositorGUI( (Compositor)node, height );
+			if ( name.Contains( '.' ) )
+			{
+				string[] tokens = name.Split( '.' );
+				return tokens[tokens.Length - 1];
+			}
+			return name;
 		}
 
-		--_indentLevel;
+		void DrawNodeWindow( int id )
+		{
+			// Get EditorNode
+			EditorData nodeData = _editorData[id];
+			TreeNode node = nodeData.node;
 
-		return height;
-	}
+			Type selectedType = CreateNodeTypeSelector( node );
+			if ( selectedType != node.GetType() )
+			{
+				ReplaceNode( nodeData, selectedType );
+			}
 
-	int CreateDecoratorGUI( Decorator node, int height )
-	{
-		if ( CreateChildrenFoldout( node, height ) )
+			if ( node is Compositor )
+			{
+				Compositor compositor = (Compositor)node;
+
+				Rect rect = new Rect( 0, 30, nodeData.nodeRect.width, 20 );
+				if ( GUI.Button( rect, "Add Child" ) )
+				{
+					TreeNode childNode = CreateNodeWithParent( typeof( NullNode ), compositor );
+					compositor._children.Add( childNode );
+				}
+			}
+
+			// Drag window last because otherwise you can't click on things
+			GUI.DragWindow();
+		}
+
+		void ReplaceNode( EditorData nodeData, Type newType )
 		{
-			Type resultType = CreateNodeTypeDropdown( node._child, height );
-			if ( node._child )
+			DeleteNode( nodeData.node );
+			CreateNodeWithExistingData( newType, nodeData );
+
+			// Update parent's reference to the node.
+			if ( nodeData.parent )
 			{
-				if ( resultType != node._child.GetType() )
+				if ( nodeData.parent is Decorator )
+				{
+					Decorator parentDecorator = (Decorator)nodeData.parent;
+					parentDecorator._child = nodeData.node;
+				}
+				else if ( nodeData.parent is Compositor )
+				{
+					Compositor parentCompositor = (Compositor)nodeData.parent;
+					parentCompositor._children[nodeData.parentIndex] = nodeData.node;
+				}
+				else
 				{
-					BehaviorTreeEditor.DeleteNode( node._child );
-					node._child = BehaviorTreeEditor.CreateNode( resultType );
+					Debug.LogError( "Parent node is neither a Decorator nor a Compositor!" );
 				}
 			}
 			else
 			{
-				node._child = BehaviorTreeEditor.CreateNode( resultType );
+				// Node is root node.
+				_behaviorTree.root = nodeData.node;
 			}
+		}
 
-			return CreateNodeGUI( node._child, ++height );
+		void DrawNodeCurve( Rect start, Rect end )
```

Comment:
did you try other shapes as well to see what looked good? Straight lines perhaps?